### ISSUE TITLE
New verb learner can only handle two argument verbs - not three argument verbs

### ISSUE THREAD

**Issue by denizbeser** [isi-vista/adam#787]

**Comment by gabbard** [comment]
See `AbstractVerbTemplateLearnerNew._candidate_templates`

**Comment by LichMaster98** [comment]
At time of writing the PR (#787) to add three-slot argument, verbs has not been merged however it is capable of examples like "Mom throws me a ball" but not "Mom throws a ball to me"

For the latter case we want to explore considering that instead of hard-coding all these patterns with their own logic, to have code to (1) generate the A/V pattern (2) then try to instantiate it using the aligned tokens. as @gabbard Suggested.

**Comment by LichMaster98** [comment]
Giving some thought to generating the A/V pattern then trying to align it to the aligned tokens gave me the following idea.

We could create a class as defined below:
```py
VerbArgumentTemplate:
    template_elements: List
    fixed_strings_mapping: Mapping(int, int)
```
Continuing with our assumption that the object aligned tokens are ordered e.g. (SLOT1, SLOT2, SLOT3) we could construct locations for the code to check if other non-aligned tokens exist via the fixed, strings mapping. The creation of the AVAVA pattern is shown below.
```py
    VerbArgumentTemplate(
        elements=[SLOT1, SLOT2, SLOT3],
        fixed_strings_mapping={
            SLOT1 : 1,
            SLOT2 : 2,
        }
    )
```

We would want to approach the next set of code thinking of going through the sentence from the last token to the start. So we'd sort our mapping by the value. We'd determine the next argument after SLOT2 then check that `middle_span.end != right_span.start` (This is from our ordered slots assumption) if true we'd then create a candidate_verb_token_span check that it's valid and if so then insert it at the given position of the `fixed_strings_mapping`. Then repeat the above for the next element in the `fixed_string_mapping.items()`. If at any point we violate a condition the entire template is abandoned - We could log the failed attempt. 

The above should also work for two-argument options as we'd just generate a single possible entry point for the fixed string. For example a VAA template could be constructed as following using a NULL_SLOT option to indicate that this is prior to slot1 -- Could also be called SLOT0?:
```py
VerbArgumentTemplate(
    elements=[SLOT1, SLOT2],
    fixed_strings_mapping={
        NULL_SLOT : 0
    }
)
```

**Comment by LichMaster98** [comment]
After discussion with @gabbard we're going to try the following approach:

Using an enum with the FixedString and SLOT tokens. Then via the `itertools.product` method we can compute multiple arrangements of possible templates: (e.g. AVA, VAA, AAV, AVAVA, etc) 

Once we've generated a template we align the recognized objects to the given slots in the template. We'll filter out the alignment which the condition of the slots being aligned in the same order as the sentence. Example: "Mom throws a ball" we'll only take the alignment which as Mom in `SLOT1` and ball in `SLOT2`. 

Then we'll go through the sentence_tokens and try to align any of the `FixedString` spots in our elements_template to the sentence. If the span is valid we'll insert it to the token elements otherwise we abandon the alignment attempt, log it, and then try again on the next template. 

For a fixed string between two arguments, we'll just select all the text which is not part of another alignment however for a fixed string before or after the arguments we'll create multiple options up to the maximum length we want to attempt to collect. 

If the template passes all checks it is converted to a SurfaceTemplateBoundToSemanticNodes.

**Comment by gabbard** [comment]
@LichMaster98 : This is correct, except that `itertools.product` is for iterating over all possible ways of filling the slots; you may be able to use `itertools.combinations` to produce the template patterns, though.

**Comment by LichMaster98** [comment]
@gabbard I'll start a branch off of the current learner integration to make this change in the verb learner.

**Comment by LichMaster98** [comment]
Closed by #799
